drivers/SmartThings/matter-switch/src/init.lua

```diff
@@ -64,25 +64,40 @@ local function initialize_switch(device)
   -- For switch devices, the profile components follow the naming convention "switch%d",
   -- with the exception of "main" being the first component. Each component will then map
   -- to the next lowest endpoint that hasn't been mapped yet.
-  for i, ep in ipairs(switch_eps) do
-    if i == 1 then
-      component_map["main"] = ep
-    else
-      component_map[string.format("switch%d", i)] = ep
+  -- Additionally, since we do not support bindings at the moment, we only want to count
+  -- On/Off clusters that have been implemented as server. This can be removed when we have
+  -- support for bindings.
+  local num_server_eps = 0
+  for _, ep in ipairs(switch_eps) do
+    if(device:supports_server_cluster(clusters.OnOff.ID, ep)) then
+      num_server_eps = num_server_eps + 1;
+      if num_server_eps == 1 then
+        component_map["main"] = ep
+      else
+        component_map[string.format("switch%d", num_server_eps)] = ep
+      end
     end
   end
 
   device:set_field(COMPONENT_TO_ENDPOINT_MAP, component_map, {persist = true})
   -- Note: This profile switching is needed because of shortcoming in the generic fingerprints
   -- where devices with multiple endpoints with the same device type cannot be detected
-  local num_switch_eps = #switch_eps
-  if num_switch_eps > 1 then
-    device:try_update_metadata({profile = string.format("switch-%d", math.min(num_switch_eps, MAX_MULTI_SWITCH_EPS))})
+  -- Also, the case where num_server_eps == 1 is a workaround for devices that have the On/Off
+  -- Light Switch device type but implement the On Off cluster as server (which is against the spec
+  -- for this device type). By default, we do not support On/Off Light Switch because by spec these
+  -- devices need bindings to work correctly (On/Off cluster is client in this case), so this device type
+  -- does not have a generic fingerprint and will join as a matter-thing. However, we have
+  -- seen some devices claim to be On/Off Light Switch device type and still implement On/Off server, so this
+  -- is a workaround for those devices.
+  if num_server_eps == 1 and device.label == "Matter Thing" then
```

Comment:
I am not sure if using the `device.label` is the best way to determine if the device was joined as a `matter-thing`. The device will join with this label so I do think it is safe to use for this case where we are checking if the device is currently joined as a matter thing so we can attempt to profile switch as needed.

However, after the switch, the device label would still be "Matter Thing", which could be somewhat confusing to a user. Is there a way to change that or is it something we have to accept with this workaround?

Reply:
> I think instead you could try to detect the shape of the matter thing profile (only refresh capability).

This is a great idea! I have implemented this with the `detect_matter_thing` function in the latest commit. Based on my testing with the matter test app as well as my nanoleaf light strip, this is working as well.